src/runtime-rs/crates/resource/src/network/network_info/network_info_from_link.rs

```diff
@@ -0,0 +1,201 @@
+        let name = &attrs.name;
+
+        Ok(Self {
+            interface: Interface {
+                device: name.clone(),
+                name: name.clone(),
+                ip_addresses: handle_addresses(handle, attrs)
+                    .await
+                    .context("handle addresses")?,
+                mtu: attrs.mtu as u64,
+                hw_addr: hw_addr.to_string(),
+                pci_addr: Default::default(),
+                field_type: link.r#type().to_string(),
+                raw_flags: attrs.flags & libc::IFF_NOARP as u32,
+            },
+            neighs: handle_neighbors(handle, attrs)
+                .await
+                .context("handle neighbours")?,
+            routes: handle_routes(handle, attrs)
+                .await
+                .context("handle routes")?,
+        })
+    }
+}
+
+async fn handle_addresses(handle: &rtnetlink::Handle, attrs: &LinkAttrs) -> Result<Vec<IPAddress>> {
+    let mut addr_msg_list = handle
+        .address()
+        .get()
+        .set_link_index_filter(attrs.index)
+        .execute();
+
+    let mut addresses = Vec::new();
+    while let Some(addr_msg) = addr_msg_list.try_next().await? {
+        if addr_msg.header.family as i32 != libc::AF_INET {
+            warn!(sl!(), "unsupported ipv6 addr. {:?}", addr_msg);
+            continue;
+        }
+        let a = Address::try_from(addr_msg).context("get addr from msg")?;
+        if a.addr.is_loopback() {
+            continue;
+        }
+
+        addresses.push(IPAddress {
+            family: if a.addr.is_ipv4() {
+                IPFamily::V4
+            } else {
+                IPFamily::V6
+            },
+            address: a.addr.to_string(),
+            mask: a.perfix_len.to_string(),
+        });
+    }
+    Ok(addresses)
+}
+
+fn generate_neigh(name: &str, n: &NeighbourMessage) -> Result<ARPNeighbor> {
+    let mut neigh = ARPNeighbor {
+        device: name.to_string(),
+        state: n.header.state as i32,
+        ..Default::default()
+    };
+    for nla in &n.nlas {
+        match nla {
+            Nla::Destination(addr) => {
+                if addr.len() != 4 {
+                    continue;
+                }
+                let dest = Ipv4Addr::new(addr[0], addr[1], addr[2], addr[3]);
+                let addr = Some(IPAddress {
+                    family: IPFamily::V4,
+                    address: dest.to_string(),
+                    mask: "".to_string(),
+                });
+                neigh.to_ip_address = addr;
+            }
+            Nla::LinkLocalAddress(addr) => {
+                if addr.len() < 6 {
+                    continue;
+                }
+                let lladdr = format!(
+                    "{:<02x}:{:<02x}:{:<02x}:{:<02x}:{:<02x}:{:<02x}",
+                    addr[0], addr[1], addr[2], addr[3], addr[4], addr[5]
+                );
+                neigh.ll_addr = lladdr;
+            }
+            _ => {}
```

Comment:
We can't log at this point [1], but should this be an error I wonder?

---

[1] - It would be kinda nice if `Result` also included a `Warn(W)` for scenarios like this to let the caller do the logging :smile:

Reply:
Thanks for your review, in this case, we get the configuration from the `nlas` list and skip the unused fields. I feel that there is no need to add warn, I'll add a comment.